3.1 release blogpost (#1284)
scala/scala-lang: blog/_posts/2021-10-19-scala-3.1.0-released.md

---
layout: blog-detail
post-type: blog
by: Paweł Marks, VirtusLab
title: Scala 3.1.0 released!
---
Hello from the Scala 3 team! It has already been six weeks since we have announced the release candidate for the first minor version after the initial release of Scala 3. Now, after two more RCs, we can confidently promote Scala 3.1.0-RC3 to a new stable release of the language. In other words, Scala 3.1.0 is officially out!

> You don't need to worry about having to migrate your projects to the new version. Scala 3 has strong compatibility guarantees, and all code working in 3.0.2 will also work in 3.1. On the other hand, switching to Scala 3.1 allows you to use many improvements and a handful of newly stabilized APIs.

Reviewer: I think "worry when upgrading your projects" would be clearer than "worry about having to migrate your projects"

Reviewer 2: I wouldn't use worry at all and maintain a more cold tone.

> Thanks to Scala 3's strong compatibility guarantees, all code 3.0.2 will work in 3.1 However ....

## What's new in 3.1

### New experimental feature: safer exceptions

A new experimental feature that allows declaring and checking which exceptions can be thrown. It relies on the effects as implicit capabilities pattern.

You can opt-in using following import.

```scala
import language.experimental.saferExceptions
```

> Now it is possible to mark types with exception that cna be thrown during the evalutaion:

Reviewer: s/cna/can/, s/evalutaion/evaluation/

```scala
def f(x: Double): Double canThrow LimitExceeded =
  if x < limit then f(x) else throw LimitExceeded())

> ```

Reviewer: extraneous right paren

You can read more in [the document proposing this feature](https://github.com/lampepfl/dotty/blob/release-3.1.0/docs/docs/reference/experimental/canthrow.md).

### Efficient bytecode for matching over strings

> When Scala 3.0 has been compiling pattern matches with `Int` as a scrutinee type, it was emitting bytecode containing very efficient instructions: `lookupswitch` or `tableswitch`. Now in Scala 3.1, similar behavior was implemented for matching over `String` literals.

Reviewer 2: ```suggestion
Scala, since 3.0.0, has been able to optimize pattern matches with `Int` as a scrutinee type by emitting bytecode containing very efficient instructions: `lookupswitch` or `tableswitch`. 
Now in Scala 3.1, similar behavior was implemented for matching over `String` literals.
```

```scala
fruit match
  case "apple"  => 1
  case "orange" => 2
  case "banana" => 3
  case _        => 0
```

> will now generate `lookupswitch` over hash codes of `String` literals instead of a chain of conditional expressions. This behavior is consistent with what Java is doing and more performant than the old approach.

Reviewer 3: ```suggestion
will now generate `lookupswitch` over hash codes of `String` literals instead of a chain of conditional expressions. This behavior is consistent with what Java does and more performant than the old approach.
```

### Support for `-Wconf` and `@nowarn`

We have added a `-Wconf` compiler flag that allows filtering and configuring compiler warnings (silence them, or turn them into errors).

> We have also integrated [the silencer plugin](https://github.com/ghik/silencer) into the compiler, which allows suppressing warnings locally using the `@nowarn` annotation.

Reviewer: (This isn't wrong, but I don't think people still think of this primarily being about silencer anymore. It's primarily a Scala 2.13 feature now. If you want to mention silencer, you could say something "as originally inspired by silencer", something like that.)

> `-Wconf` and `@nowarn` work largely the same way as in Scala 2, but some filters for selecting warnings are different. For the more details see the output of `-Wconf:help` flag.

Reviewer: That output is rather lengthy. To do a detailed comparison of the Scala 2 output with the Scala 3 output would be difficult — unless you're unusually eagle-eyed, you'd have to pipe them through `diff`. It might be better to just actually tell the reader what the differences are?

### Simplified Manifest synthesis

For the sake of compatibility with Scala 2, the Scala 3 compiler now generates given instances for the `Manifest` trait. The new implementation is a slightly simplified approximation of the Scala 2 implementation. It guarantees that any of the expressions:

- `manifest[A] == manifest[B]`
- `manifest[A].runtimeClass == manifest[B].runtimeClass`
- `optManifest[A] == optManifest[B]`
- `optManifest[A].asInstanceOf[ClassTag[A]].runtimeClass == optManifest[B].asInstanceOf[ClassTag[B]].runtimeClass`

> that was true in Scala 2, will also remain true when compiled with Scala 3.

Reviewer: s/that was/that were/

> ### Other changes

Reviewer 4: Mention the upgrade to Scala.js 1.7.x here? It impacts users since they will need to upgrade their Scala.js version to 1.7.0+ before they can upgrade to Scala 3.1.

Author: Can you please suggest the change? I don't know much about Scala.js and don't want to include anything that can be misleading for the users.

> - `Unapply.apply` constructor was added to the reflection API

Reviewer: What's the purpose of this?

> - Scastie was integrated into Scaladoc to make snippets interactive.

Reviewer 3: Maybe add a link to some place in the docs that people could just click to see how this works?

Author: @pikinier20

> - `@experimental` spec was changed. See more detail in [the design document](https://github.com/lampepfl/dotty/blob/release-3.1.0/docs/docs/reference/other-new-features/experimental-defs.md)

Reviewer: It's always better to summarize the changes briefly before just dumping the reader directly into the doc, IMO.

- Now `TastyInspector.{inspectTastyFiles, inspectTastyFilesInJar, inspectAllTastyFiles}` return a boolean value indicating whether the process succeeded
- A `Wildcard` was made a subtype of `Ident` in the reflection API

> - `TypedOrTest` was added as a super type of `Typed` in the reflection API

Reviewer: s/super type/supertype/

> Beside that scala 3.1.0 introduced multiple small improvements and fixed handful of bugs. You can see [the detailed changelog](https://github.com/lampepfl/dotty/releases/tag/3.1.0) on GitHub.

Reviewer: s/scala/Scala/, s/fixed handful/fixed a handful/

## What's next

> During the Scala 3.1.0 stabilization period, which took the last six weeks, we haven't stopped improving the language and fixing the bugs in the compiler. You can already test the results of our work, as they are released as Scala 3.1.1-RC1. [The full changelog](https://github.com/lampepfl/dotty/releases/tag/3.1.1-RC1) is as always available on GitHub.

Reviewer 2: ```suggestion
During the Scala 3.1.0 stabilization period, which took the last six weeks, we haven't stopped improving the language and fixing the in the compiler. You can already test the results of our work, using the published release candidates: Scala 3.1.1-RC1. [The full changelog](https://github.com/lampepfl/dotty/releases/tag/3.1.1-RC1) is as always available on GitHub.
```

We have added a `-Wconf` compiler flag that allows filtering and configuring compiler warnings (silence them, or turn them into errors).

We have also brought back `@nowarn` annotation from Scala 2.13, originally inspired by [the silencer plugin](https://github.com/ghik/silencer). It can be used for suppressing warnings directly in the code, in the places where they are expected to occur.

`-Wconf` and `@nowarn` work largely the same way as in Scala 2, but some filters for selecting warnings are different. For the more details see the output of `-Wconf:help` flag.

### Simplified Manifest synthesis

For the sake of compatibility with Scala 2, the Scala 3 compiler now generates given instances for the `Manifest` trait. The new implementation is a slightly simplified approximation of the Scala 2 implementation. It guarantees that any of the expressions:

- `manifest[A] == manifest[B]`
- `manifest[A].runtimeClass == manifest[B].runtimeClass`
- `optManifest[A] == optManifest[B]`
- `optManifest[A].asInstanceOf[ClassTag[A]].runtimeClass == optManifest[B].asInstanceOf[ClassTag[B]].runtimeClass`

that were true in Scala 2, will also remain true when compiled with Scala 3.

> 

Reviewer 5: I think that https://github.com/lampepfl/dotty/pull/13519 "Raise type error on unreducible match type" is very important to include

- Scala 3.1 is backward source compatible: any code that was working in 3.0 will also work in 3.1.
- Scala 3.1 is backward binary compatible: you can use dependencies compiled with Scala 3.0 in 3.1 projects.
- Scala 3.1 is _not_ forward binary compatible: you _cannot_ use dependencies compiled with Scala 3.1 in Scala 3.0 projects.
- Updating Scala to 3.1 is not forward compatible change. If you are a library maintainer you should only do that in minor release of your project.

## What's new in 3.1

### New experimental feature: safer exceptions

A new experimental feature that allows declaring and checking which exceptions can be thrown. It relies on the effects as implicit capabilities pattern.

You can opt-in using following import.

```scala
import language.experimental.saferExceptions
```

Now it is possible to mark types with exception that can be thrown during the evaluation:

```scala
def f(x: Double): Double canThrow LimitExceeded =
  if x < limit then f(x) else throw LimitExceeded()
```

You can read more in [the document proposing this feature](https://github.com/lampepfl/dotty/blob/release-3.1.0/docs/docs/reference/experimental/canthrow.md).

> ### Efficient bytecode for matching over strings

Reviewer 5: would be very great to have a section for hierarchical mirror generation: https://github.com/lampepfl/dotty/pull/11686 e.g. 

### Summon Mirrors for Hierarchical Sealed Types
Consider the following sealed hierarchy:
```scala
sealed trait Top
case class Middle() extends Top with Bottom
sealed trait Bottom extends Top
```
Previously you would not be able to summon an instance of
`scala.deriving.Mirror.SumOf` for `Top`, as its child `Bottom`
is also sealed. In Scala 3.1 this is now possible, enabling
compiletime reflection over nested hierarchies of sealed types.

The only change necessary for users is that you should now
consider that one of the `MirroredElemTypes` for a `Mirror.SumOf`
could also be a sum type, not just product types.

---
Hello from the Scala 3 team! It has already been six weeks since we have announced the release candidate for the first minor version after the initial release of Scala 3. Now, after two more RCs, we can confidently promote Scala 3.1.0-RC3 to a new stable release of the language. In other words, Scala 3.1.0 is officially out!

## Compatibility notcie

This is a first *minor* release after the initial release of Scala 3.0. This has following consequences:

- Scala 3.1 is backward source compatible: any code that was working in 3.0 will also work in 3.1.
- Scala 3.1 is backward binary compatible: you can use dependencies compiled with Scala 3.0 in 3.1 projects.
- Scala 3.1 is _not_ forward binary compatible: you _cannot_ use dependencies compiled with Scala 3.1 in Scala 3.0 projects.

> - Updating Scala to 3.1 is not forward compatible change. If you are a library maintainer you should only do that in minor release of your project.

Reviewer 6: This seems to be making some specific assumptions about what minor releases of projects should contain and the Scala.js release notes don't say anything like that /cc @sjrd

Reviewer 7: Actually I disagree with your comments. A library that follows semver cannot update to `3.1` in a patch release.

Reviewer 4: > Actually I disagree with your comments. A library that follows semver cannot update to `3.1` in a patch release.

Irrespective of that, this is not the place to assume that library authors are using semver, or any other scheme for that matter. They need to do the right thing for their own versioning scheme.

Reviewer: s/many/may/

This is a first *minor* release after the initial release of Scala 3.0. This has following consequences:

- Scala 3.1 is backward binary compatible: you can use dependencies compiled with Scala 3.0 in 3.1 projects.
- Scala 3.1 is _not_ forward binary compatible: you _cannot_ use dependencies compiled with Scala 3.1 in Scala 3.0 projects.
- Updating Scala to 3.1 is not a forward compatible change. If you are a library maintainer and are following SemVer you may want to update in your next minor release.

Although we cannot guarantee full source compatibility between minor versions, we have put a lot of effort into assuring that all code that was working in 3.0.2, except some rare cases, will also work in 3.1.0. This means that if you are an application developer, you can confidently update the compiler version to take advantage of the newest improvements.

> If you are a library maintainer, updating to 3.1.0 will force all of your users to need to update as well.  You may consider still publishing your library using 3.0.2 to ensure that users who are reluctant to update a compiler can still use it. On the other hand, we still encourage you to test your library using 3.1.0. We understand that the current state of binary compatibility may be unsatisfactory for library maintainers. We are actively working on improvements in this area in future releases on scala.

Reviewer: s/on scala/of Scala/
